git2-hyper/src/lib.rs

```diff
@@ -0,0 +1,270 @@
+        let parsed = Uri::from_str(&url).map_err(|_| self.err("invalid url, failed to parse"))?;
+        let host = match parsed.host() {
+            Some(host) => host,
+            None => return Err(self.err("invalid url, did not have a host")),
+        };
+
+        // Prep the request
+        debug!("request to {}", url);
+        let client = self.handle.lock().unwrap();
+
+        let method =
+            Method::from_bytes(self.method.as_bytes()).map_err(|_| self.err("invalid method"))?;
+        let request = Request::builder()
+            .method(method)
+            .uri(&url)
+            .header(header::USER_AGENT, agent)
+            .header(header::HOST, host)
+            .header(header::EXPECT, "");
+
+        let request = if data.is_empty() {
+            request.header(header::ACCEPT, "*/*")
+        } else {
+            request
+                .header(
+                    header::ACCEPT,
+                    format!("application/x-git-{}-result", self.service),
+                )
+                .header(
+                    header::CONTENT_TYPE,
+                    format!("application/x-git-{}-request", self.service),
+                )
+        };
+
+        let request = request
+            .body(Body::from(Vec::from(data)))
+            .map_err(|_| self.err("invalid body"))?;
+
+        let mut res = rt.block_on(client.request(request)).unwrap();
+        let headers = res.headers();
+
+        let content_type = match headers.get(header::CONTENT_TYPE) {
+            Some(v) => Some(v.to_str().unwrap()),
+            None => None,
+        };
+
+        let code = res.status();
+        if code.as_u16() != 200 {
+            return Err(self.err(
+                &format!(
+                    "failed to receive HTTP 200 response: \
+                     got {}",
+                    code
+                )[..],
+            ));
+        }
+
+        // Check returned headers
+        let expected = match self.method {
+            "GET" => format!("application/x-git-{}-advertisement", self.service),
+            _ => format!("application/x-git-{}-result", self.service),
+        };
+
+        if let Some(content_type) = content_type {
+            if content_type != expected {
+                return Err(self.err(
+                    &format!(
+                        "expected a Content-Type header \
+                         with `{}` but found `{}`",
+                        expected, content_type
+                    )[..],
+                ));
+            }
+        } else {
+            return Err(self.err(
+                &format!(
+                    "expected a Content-Type header \
+                         with `{}` but didn't find one",
+                    expected
+                )[..],
+            ));
+        }
+
+        // Ok, time to read off some data.
+        let body = rt.block_on(res.body_mut().data());
```

Comment:
I’m not really familiar with hyper, but iiuc this will read the entire response body into memory, which is not so desirable. From the docs, I think you’ll want to execute the request here, and then store the response such that you can poll the body stream for the next chunk of data in the `Read` impl.

Reply:
fixed in 8854f95